doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,384 @@
+| `SELECT`  | Allows reading rows from an object.                                      | r(”read”)    | Table, View, Materialized View, Source        | Yes             |
+| `INSERT`  | Allows inserting into an object.                                         | a(”append”)  | Table                                         | Yes             |
+| `UPDATE`  | Allows updating an object (requires SELECT if a read is necessary).      | w(”write”)   | Table                                         | Yes             |
+| `DELETE`  | Allows deleting from an object (requires SELECT if a read is necessary). | d            | Table                                         | Yes             |
+| `CREATE`  | Allows creating a new object within another object.                      | C            | Database, Schema, Cluster                     | Yes             |
+| `USAGE`   | Allows using an object or looking up members of an object.               | U            | Database, Schema, Connection, Secret, Cluster | Yes             |
+
+We will support the following object types:
+
+| Object Type          | All Privileges | From PostgreSQL |
+|----------------------|----------------|-----------------|
+| `DATABASE`           | UC             | Yes             |
+| `SCHEMA`             | UC             | Yes             |
+| `TABLE`              | arwd           | Yes             |
+| `VIEW`               | r              | Yes             |
+| `MATERIALIZED  VIEW` | r              | Yes             |
+| `INDEX`              |                | Yes             |
+| `SOURCE`             | r              | No              |
+| `SINK`               |                | No              |
+| `CONNECTION`         | U              | No              |
+| `SECRET`             | U              | No              |
+| `CLUSTER`            | UC             | No              |
+
+When any of these objects are created, the creating role is assigned as the owner of that object. Only the owner (and
+superusers) can destroy or alter that object. The owner is given all privileges on an object by default, though these
+privileges can be revoked.
+
+PostgreSQL allows arwd privileges on all table like objects (view, materialized view, etc.) even though they aren't
+useful. We remove privileges that don't make sense.
+
+Below is a summary of the default owners and privileges of all builtin objects:
+
+- The `mz_system` cluster will be owned by the `mz_system` role.
+- The `mz_system` role will have `UC` privileges on the `mz_system` cluster.
+- The `mz_introspection` cluster will be owned by the `mz_introspection` role.
+- The `mz_introspection` role will have `UC` privileges on the `mz_introspection` cluster.
+- All roles will have `U` privileges on the `mz_introspection` cluster.
+- The `default` cluster will be owned by the `mz_system` role.
```

Comment:
> > the first user invited to an org is always a superuser
> 
> Just to clarify, this logic is done in the frontend right? The adapter will not care about the first user vs the nth user, it will just look at the Frontegg roles to determine who is a superuser.


Yep, it's all handled by the Frontegg signup process.